### ISSUE TITLE
[proposal] multi language support ; language neutrality ; hstore for all osm name ;  simple config   

### ISSUE THREAD

**Issue by ImreSamu** [openmaptiles/openmaptiles#80]
### Problem:

* Which languages supported and why ?  
* language neutrality 

### Proposal 

So I am thinking about multi language support .. 

in the future I would like to Experimenting  this plan  :  

####  Step 1
- [ ] add  `hstore_keep`  mapping  to imposm3 ( upstream ? )
   *   https://github.com/ImreSamu/imposm3_fieldtypes/blob/master/mapping/fieldtype_hstore_tags_keep.go

and store all openstreetmap name:*  tags 
 *  http://taginfo.openstreetmap.org/search?q=name%3A
 *  mapping :  
```yaml
-   name: name_all
    key: null
    type: hstore_tags_keep
    args:
        keep_keys:
             name:*
```
####  Step 2
- [ ] add `name_all::hstore`   field to layers

####  Step 3
- [ ] add requested languages for the configurations 
```yaml
tileset:
  requested_languages:  
    - en
    - hu
    - de
    - es
  layers:
    - layers/boundary/boundary.yaml
    - layers/place/place.yaml
    - layers/poi/poi.yaml
  name: OpenMapTiles v1.0
  description: "A tileset showcasing all layers in OpenMapTiles. http://openmaptiles.org"
  attribution: "<a href=\"http://www.openstreetmap.org/about/\" target=\"_blank\">&copy; OpenStreetMap contributors</a>"
  center: [-12.2168, 28.6135, 4]
```

####  Step 4
- [ ]  generate to final query for  tm2source ( by `openmaptiles-tools` based on `requested_languages`  )
```sql
query: (SELECT geometry, name 
           ---  requested_languages : 
           ,name_all->'name:en' as name_en
           ,name_all->'name:hu' as name_hu
           ,name_all->'name:de' as name_de
           ,name_all->'name:es' as name_es
           ----------------------------------------------
          , class, rank FROM layer_place(!bbox!, z(!scale_denominator!), !pixel_width!)) AS t
```

#### Summary 
if the users want more or less languages , they need modify 1 simple config file ..
```
  requested_languages:  
    - en
    - hu
    - de
    - es
```

### Edit1    ( based on comments ) 
* other similar fields ( based on Mapzen ) :  
  *  `alt_name:*`  ->  `requested_alt_languages`
  *  `old_name:*` ->  `requested_old_languages`

**Comment by ImreSamu** [comment]
We need special new  imposm3 mappings ...   
and with https://github.com/openmaptiles/openmaptiles/issues/73  it will be slower to implement in the upstream.

So I moved this to the https://github.com/openmaptiles/openmaptiles/milestone/8  ;

**Comment by nyurik** [comment]
I would suggest going a somewhat different route:  store all languages in the same PBF tile with geometries. For server-side rendering, add an additional "loading" layer which will decode stored PBF, and on the fly substitute the "name" field that contains an object (lang->name) with a single name string. This way all existing tools, like mapnik, can continue working without any changes. For the client side rendering, it could either also use the same server side re-packing, or it can implement its own language picking.  I plan to start working on this "pbf repacker" fairly soon.  Also, this means that there has to be a "storage repacker" - mapnik is not able to deal with hstore, but we could fool it by generating a JSON string in SQL with all available languages, and after it passes through mapnik, have a PBF re-encoder - which will convert JSON->named values.

See also MVT discussion https://github.com/mapbox/vector-tile-spec/issues/75#issuecomment-249655916

**Comment by jirik** [comment]
`name_de` implemented: https://github.com/openmaptiles/openmaptiles/pull/190

**Comment by klokan** [comment]
@nyurik could you please gives us update on the status on Wikimedia maps?

- How have you solved the storage of all different languages in MVT in the end?
- Where is the latest version of maps.wikimedia.org - can we follow on GitHub?
- What about to meet shortly on Hangout to synchronise the progress again?

**Comment by nyurik** [comment]
I have some if this stuff run on my machine, but still a bit raw.  Let's sync up on Friday? I'll PM

**Comment by nyurik** [comment]
I have some if this stuff run on my machine, but still a bit raw.  Let's sync up on Friday? I'll PM

**Comment by nyurik** [comment]
@klokan sent an invite.  Current status:
* The pbfs from [new tiles](http://shorter.wmflabs.org) (can get as .pbf too) contain a field with all names as a single JSON-encoded value. Not very efficient, but it is a needed step
* I have a tilelive source that can modify a PBF on the fly from JSON->multiple name:*=... values, as well as modify it again from multiple into a single "name" = ...  based on language fallback rules. See https://github.com/kartotherian/kartotherian-babel

**Comment by stirringhalo** [comment]
Any cool discussions come from this?

**Comment by nyurik** [comment]
@stirringhalo see right above - i'm experimenting with my version. I think openmaptiles is working on a limited subset of languages instead.

**Comment by klokan** [comment]
Our next steps related to multi-lingual vector tiles are proposed at #211

**Comment by nyurik** [comment]
I'm still unclear on the fallback strategy. We should clarify the fallback matrix. It may depend on the location of the object (should an object in China use the same rules as the one in Canada?), and how we should use other related tags like name and name_int. Also, we may consider [traansliteration](https://www.npmjs.com/package/transliteration) library for some cases.

* Show the map in English

| Which tags are set on an object | in USA | in China |
|---|---|---|
|name, name_en, name_int| ?? | ?? |
|name, name_zh, name_int| ?? | ?? |
|name, name_en| ?? | ?? |

* Show the map in Chinese

| Which tags are set on an object | in USA | in China |
|---|---|---|
|name, name_en, name_int| ?? | ?? |
|name, name_zh, name_int| ?? | ?? |
|name, name_en| ?? | ?? |

**Comment by nyurik** [comment]
I wrote a helper library https://github.com/kartotherian/language-scripts - it converts a language code to a script code  ("en" -> "Latn") - based on CLDR data.  I think it can be used for some generic fallback strategies like  "any latin-based before any non-latin based" sorting.

**Comment by klokan** [comment]
Multilingual tiles are implemented in #211
